phoenix-core/src/it/java/org/apache/phoenix/end2end/index/UncoveredGlobalIndexRegionScannerIT.java

```diff
@@ -369,18 +523,97 @@ public void testCount() throws Exception {
         }
     }
 
+    @Test
+    public void testFailDataTableRowUpdate() throws Exception {
+        String dataTableName = generateUniqueName();
+        populateTable(dataTableName); // with two rows ('a', 'ab', 'abc', 'abcd') and ('b', 'bc', 'bcd', 'bcde')
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            String indexTableName = generateUniqueName();
+            conn.createStatement().execute("CREATE " + (uncovered ? "UNCOVERED " : " ") + "INDEX "
+                    + indexTableName + " on " + dataTableName + " (val1)");
+            // Configure IndexRegionObserver to fail the data table update
+            // and check that this does not impact the correctness
+            IndexRegionObserver.setFailDataTableUpdatesForTesting(true);
+            conn.createStatement().execute("upsert into " + dataTableName + " (id, val2) values ('a', 'abcc')");
+            try {
+                conn.commit();
+                fail();
+            } catch (Exception e) {
+                // this is expected
+            }
+            IndexRegionObserver.setFailDataTableUpdatesForTesting(false);
+
+            conn.createStatement().execute("upsert into " + dataTableName + " (id, val3) values ('a', 'abcdd')");
+            conn.commit();
+            String selectSql = "SELECT"+ (uncovered ? " " : "/*+ INDEX(" + dataTableName + " "
+                    + indexTableName + ")*/ ") + "val2, val3 from " + dataTableName
+                    + " WHERE val1  = 'ab'";
+            // Verify that we will read from the first index table
+            assertExplainPlan(conn, selectSql, dataTableName, indexTableName);
+            ResultSet rs = conn.createStatement().executeQuery(selectSql);
+            assertTrue(rs.next());
+            assertEquals("abc", rs.getString(1));
+            assertEquals("abcdd", rs.getString(2));
+            assertFalse(rs.next());
+        }
+    }
+    @Test
+    public void testFailPostIndexDeleteUpdate() throws Exception {
+        String dataTableName = generateUniqueName();
+        populateTable(dataTableName); // with two rows ('a', 'ab', 'abc', 'abcd') and ('b', 'bc', 'bcd', 'bcde')
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            String indexTableName = generateUniqueName();
+            conn.createStatement().execute("CREATE " + (uncovered ? "UNCOVERED " : " ") + "INDEX "
+                    + indexTableName + " on " + dataTableName + " (val1)");
+            String selectSql = "SELECT id from " + dataTableName + " WHERE val1  = 'ab'";
+
+            // Verify that we will read from the index table
+            assertExplainPlan(conn, selectSql, dataTableName, indexTableName);
+            ResultSet rs = conn.createStatement().executeQuery(selectSql);
+            assertTrue(rs.next());
+            assertEquals("a", rs.getString(1));
+            assertFalse(rs.next());
+            TestUtil.dumpTable(conn, TableName.valueOf(indexTableName));
```

Comment:
nit: are these TestUtil.dumpTable calls still necessary?

Reply:
Good catch. Those were added for debugging. I will remove them.